LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

veronikaslc opened the pull request.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

OrdiNeu commented:
It is possible to select multiple options when `maxAnswers=1` (visual glitch only, the last-selected option appears to be used)
- In the test runmode, install HANCESTRO
- Create a Multiple Choice Test form.
- Enter `as` in the textbox for the first question, and then click the info box next to East Asian
- You can select multiple options (this appears to just be a visual glitch)
![image](https://user-images.githubusercontent.com/4656440/117194443-1b601e00-adb2-11eb-939e-848b2c136c48.png)

Previously selected options do not appear selected in the checkboxes (in any set of maxAnswers).
- In the test runmode, install HANCESTRO
- Create a Multiple Choice Test form.
- Select European
- Enter `eu` in the first textbox, and then click the info box next to European American
- European does not appear selected.
![image](https://user-images.githubusercontent.com/4656440/117195054-d1c40300-adb2-11eb-8c2e-54bf4a7edba4.png)

The vocabulary filter does not seem to work with this change.
- In the test runmode, install HANCESTRO
- On the Dashboard, add a new filter for a Race/Ethnicity Reported question
- Use the browser to select something. It shows up as a radio box (works) but selecting anything does not affect the filter.
![image](https://user-images.githubusercontent.com/4656440/117195521-5b73d080-adb3-11eb-9dc6-8a0d60ebed31.png)

sashaandjic commented:
I can no longer select Ethnicity as shown in the screenshot:
![image](https://user-images.githubusercontent.com/23243244/117196487-89a5e000-adb4-11eb-8b47-dcef52aa58eb.png)

After successful selection from vocabulary browser, my original text does not clear.
![image](https://user-images.githubusercontent.com/23243244/117196774-e99c8680-adb4-11eb-862e-c8e294d3a06c.png)

veronikaslc commented:
Last commit fixes last reported bug.

marta- commented:
> The vocabulary filter does not seem to work with this change.

Since this task is about selection on the form, we don't need to make the filters work with this change as part of this PR, **but** we should not be displaying the radiobuttons when browsing the ontology from the filters. (Also there should be no radiobuttons/checkboxes when browsing the vocabulary in the Vocabularies admin page #626 ).

In addition to @OrdiNeu 's comments:
* IMO the selection should be changed as soon as a checkbox/radio is clicked, not after closing the browser
* It would probably look better if the radio/checkbox is after the expand/collapse button. LFS-1136 (#629) should be integrated to get a better idea about the final UI.

veronikaslc commented:
Work in progress. Rebased on  LFS-1115 PR for further development.

sashaandjic commented:
White screen of death after clicking on LEARN MORE button 
![image](https://user-images.githubusercontent.com/23243244/117418188-210c4f80-aee9-11eb-9c83-188d24bd71cb.png)

veronikaslc commented:
Work in progress means the work is still in progress ans PR should not be retested.

veronikaslc commented:
Work in progress

veronikaslc commented:
@sashaandjic Ready for test. Note - selecting options will **only work for Forms, not for filters.**

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +226,21 @@ function VocabularyBranch(props) {
+        {/* Browser term select tools */}
+	    { addCheckbox &&
+	      <Checkbox
+	        color="primary"
```
IMO the color of the browser checkboxes and radios should be consistent with the form controls. In the form we have orenge-red (secondary color) while in the browser we have midnight blue (primary color). I suggest changing it to secondary, wdyt? @sashaandjic @veronikaslc 
```suggestion
	        color="secondary"
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +226,21 @@ function VocabularyBranch(props) {
+	    { addRadio &&
+	       <Radio
+	         checked={selectedPaths.includes(path)}
+	         color="primary"
```
```suggestion
	         color="secondary"
```

marta- reviewed (dismissed):
The selection status in the vocabulary browser sometimes appears to be one step behind the form, [see screencast](https://user-images.githubusercontent.com/651980/119162607-923f1d00-ba28-11eb-9222-6ab947b232d7.gif).

sashaandjic commented:
Problem found:
1. type "Bri" in the `Race/Ethnicity reported` field of `Patient Information` form
2. click on `"i"` next to `British`
3. click on `LEARN MORE`
4. select first `British`, then `European `option---> `Caucasian `(a synonym for European) and `British `will be selected in the form as expected
5. click on `"X"` to close the pop up
![image](https://user-images.githubusercontent.com/23243244/119611816-2e6e7880-bdc9-11eb-95e5-cd80b3529e97.png)
6. "Bri" should be cleared up upon closing, but it is still there along with the 2 options that initially popped up
![image](https://user-images.githubusercontent.com/23243244/119610623-86a47b00-bdc7-11eb-9a32-0bed875935e5.png)

Another problem:
- create a new `Patient Information` form
- repeat steps: 1, 2 and 3
- first select `European `-->  `Caucasian `(a synonym for European) will be added to the form as expected
- select `British` --> `British `gets added to the form (**correct**), but `Caucasian `option gets removed (**wrong**)
- close popup
- "Bri" is still there with the 2 initial options

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Make form controls consistent

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Synchronize interactions with answers list and vocabulary tree

sashaandjic commented:
Should these 4 root options be allowed for selection?
![image](https://user-images.githubusercontent.com/23243244/119941606-d36e8a00-bf5e-11eb-9a36-aab4cf2e21e0.png)

sashaandjic commented:
This problem is still present:
- type "Bri" in the Race/Ethnicity reported field of Patient Information form
- click on "i" next to British
- click on `LEARN MORE`
- first select European --> Caucasian (a synonym for European) will be added to the form as expected
- select British --> British gets added to the form (**correct**), but Caucasian option gets removed (**wrong**)

sashaandjic commented:
Unable to make selection other than predefined when MaxAnswers=1

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Synchronize selecting default options

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Do not clean input after selection made for questions with MaxAnswers=1; Disable event propagation to prevent onTermClick events when selecting answers; Fixed warning of dev as p decendent

sashaandjic commented:
Still unable to make selection when `maxAnswers=1` unless I press `i` and 'LEAR MORE' buttons. 
Please note once I made a selection, I have to close 3 pop ups manually.

https://user-images.githubusercontent.com/23243244/120046938-4f111b00-bfe1-11eb-9005-3dac3a5a32f3.mp4

sashaandjic commented:
British-Europian combo is still not behaving properly.
- type "bri"
- click on `i` next to British option
- click on `LEARN MORE`
- select European --> Caucasian gets selected in the background (correct)
- select British  --> British gets selected in the background (correct), but Caucasian gets de-selected (wrong)

Other scenario:
- type "bri"
- click on `i` next to British option
- click on `LEARN MORE`
- select British, then European
- close pop ups --> all good: British and Caucasian are selected
- type "bri" again
- click on `i` next to British option
- click on `LEARN MORE`
- only "British" is selected (instead of "British" and "European")

veronikaslc commented:
>Please note once I made a selection, I have to close 3 pop ups manually.
Clickaway works for dropdown and info windows. Only vocabulary tree has to be closed manually - was requested by @sdumitriu not to lose the tree navigation results on accidental click.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Select answer from a dropdown on click for questions with maxAnswers= 1

sashaandjic commented:
When maxAnswers=1, once selection is made the input does not get cleared. Please see the videos when maxAnswers=0 and maxAnswers>0  for comparison:


https://user-images.githubusercontent.com/23243244/120083071-4d545f80-c094-11eb-8600-28a8cb1c14b8.mp4


https://user-images.githubusercontent.com/23243244/120083067-3f9eda00-c094-11eb-9b25-85dc3bb80d65.mp4

SUGGESTION: since we currently do not have vocabulary fields with maxAnswers=1 condition, I suggest we log and fix this problem separately.

sdumitriu commented:
> **SUGGESTION**: Since we currently do not have vocabulary fields with maxAnswers=1 condition, I suggest we log and fix this problem separately.

I think we do, in one of the forms starting with T (Tumor or TissueMetrix).

veronikaslc commented:
@sashaandjic Work in progress.  I had only fixed the magnifying search glass behavior with last commit. Please do not test PRs until it has "Test me!" label.

veronikaslc commented:
Bad behaviour, should be fixed here. Working on it.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Clean input on selecting predefined answer for questions with maxAnswers=1

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Added more checks for existing of answers

sdumitriu reviewed (changes requested):
LFS mode, new Patient Information form, search for something in the Race question, more info, browse. Go up to a term that has many children, then select several of these children. Close the vocabulary browser and term info dialogs. All the selected terms appear selected. Save, close, open again -> only one of the terms was actually saved.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Work in progress

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Added chips for selected terms, some refactorings

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +226,22 @@ function VocabularyBranch(props) {
+	      <Checkbox
+	        color="secondary"
+	        checked={selectedPaths.includes(path)}
+	        onClick={(evt) => {event.stopPropagation(); onTermSelect(evt, path, name);}}
```
Who is `event`, shouldn't it have been `evt.stopPropagation()`? But isn't it redundant with the first line of `onTermSelect`? I would also rename the function to `onSelectionChanged` to cover both select and unselect, and to simplify the code I would remove the path and name params which should be already available in the scope of the function.

```suggestion
	        onClick={onSelectionChanged}
```

marta- replied:
Addressed in 1b7e199

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -60,6 +66,7 @@ function VocabularyBranch(props) {
   const [ childrenData, setChildrenData ] = useState();
   const [ children, setChildren ] = useState([]);
   const [ expanded, setExpanded ] = useState(defaultOpen);
+  const [ selectedPaths, setSelectedPaths] = useState(initialSelection ? initialSelection.map(item => item[1]) : []);
```
`import { LABEL_POS, VALUE_POS } from "../questionnaire/Answer"` instead of hardcoding the positions. Technically the index is correct here, but we've had annoying bugs in the past that stemmed from swapping the indices.

marta- replied:
Also, can we `filter` in the `VocabularyTree` component and pass along only the paths to the branch, instead of doing this processing in each `VocabularyBranch`?

marta- replied:
Done in 1b7e199

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -145,14 +182,20 @@ function VocabularyTree(props) {
+      { allowTermSelection &&
+        <div className={classes.selectionContainer}>
+          <Typography variant="body2" component="span">{questionText}:</Typography>
+          { selectedTerms?.filter(i => i[0]).map(s => <Chip key={s[1]} variant="outlined" size="small" label={s[0]} onDelete={() => onRemoveOption(s[0], s[1])} color="primary" className={classes.selectionChips}/>) }
```
```suggestion
          { selectedTerms?.filter(i => i[0]).map(s => <Chip key={s[1]} variant="outlined" size="small" label={s[0]} onDelete={() => onRemoveOption(...s)} color="primary" className={classes.selectionChips}/>) }
```

marta- replied:
Addressed in 1b7e199

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -108,7 +108,13 @@ const thesaurusStyle = theme => ({
+  },
+  selectionChips: {
+    margin: theme.spacing(0.5, 0.5),
+  },
```
Let's also add some styling for the checkboxes and radios, as the alignment of the tree is a bit off ATM. Setting `margin: theme.spacing(-.25, 0, 0, 1.5)` seems to improve it.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -359,6 +373,11 @@ function VocabularyQuery(props) {
+          allowTermSelection={allowTermSelection}
+          initialSelection={initialSelection}
+          onCloseBrowser={onCloseBrowser}
+          questionText={questionDefinition?.text}
```
I would pass the whole `questionDefinition` as one prop instead of `maxAnswers` and `text` separately. The number of props for this component is becoming overwhelming.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -277,6 +285,12 @@ function VocabularyQuery(props) {
   }
 
+  let onCloseBrowser = (selectedTerms, removedTerms) => {
+    selectedTerms.map(item => onClick(item[1], item[0]));
+    removedTerms.map(item => removeOption(item[1], item[0]));
```
Why first [1], then [0]? `selectedTerms` and `removedTerms` were introduced in this PR, can they be structured consistently with the VocabularyQuestion?

veronikaslc replied:
They are. We are getting the selected values from VocabularyQuestion via `initialSelection` in the format [label, path], but call add/remove on them in order (path, label) in MultipleChoice.

marta- replied:
Right, they come that way from `MultipleChoice`.

What really needs to be done is `import { LABEL_POS, VALUE_POS } from "../questionnaire/Answer";` and use them instead of hardcoded [0], [1] to make the code easier to understand.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -135,6 +167,11 @@ function VocabularyTree(props) {
+        addRadio={addRadio}
+        addOption={onAddOption}
+        removeOption={onRemoveOption}
+        initialSelection={selectedTerms}
```
The terminology is confusing. `selectedTerms` is supposed to be updated in real time by branches, so the better name for this prop is `currentSelection`.

Note: it should remain as `initialSelection` in all other components (quety, browser, tree), as only `VocabularyBranch` has this behavior.

marta- replied:
Done in 1b7e199

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -117,6 +127,28 @@ function VocabularyTree(props) {
+      let newRTerms = removedTerms.slice();
+      newRTerms.push([name, path]);
+      setRemovedTerms(newRTerms);
+    }
```
I'm not sure I follow the logic here. The comment says "they were selected before, add to removed". When is "before"? should the removed term be checked against `initialSelection` instead? The following seems to make more sense to me, but maybe I'm missing something.

```suggestion
    let newTerms = selectedTerms.filter(item => item[1] != path);
    setSelectedTerms(newTerms);
    if (initialSelection.some(item => item[1] == path)) {
      // they were selected before, add to removed
      let newRTerms = removedTerms.slice();
      newRTerms.push([name, path]);
      setRemovedTerms(newRTerms);
    }
```

marta- replied:
Addressed in 96823e0

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -117,6 +127,28 @@ function VocabularyTree(props) {
+    setSelectedTerms(newTerms);
+    // remove from removed
+    let newRTerms = removedTerms.filter(item => item[1] != path);
+    setRemovedTerms(newRTerms);
```
These changes should be atomic:
```
setSelectedTerms(old => {
  ...
  return new;
});
```
otherwise one branch might end up updating the old state. I believe this is the cause of the bug where branches displayed together appear to replace each other's selection rather than appending.

marta- replied:
Fixed in 96823e0

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +226,22 @@ function VocabularyBranch(props) {
+                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}
+                  component="div">
+        {/* Browser term select tools */}
+	    { addCheckbox &&
+	      <Checkbox
+	        color="secondary"
+	        checked={selectedPaths.includes(path)}
+	        onClick={(evt) => {event.stopPropagation(); onTermSelect(evt, path, name);}}
+	      /> }
+	    { addRadio &&
+	       <Radio
+	         checked={selectedPaths.includes(path)}
+	         color="secondary"
+	         onChange={(evt) => {event.stopPropagation(); onTermSelect(evt, path, name);}}
+	       /> }
```
Could we use the same event (e.g. `onChange`) for both `Checkbox` and `Radio`, and pass a single prop named `selectionComponent` instead of the 2 props` addCheckbox`/`addRadio` which are mutually exclusive?

An example of passing components as props is in [VocabularyQuestion](https://github.com/ccmbioinfo/lfs/blob/dev/modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx).

marta- replied:
This is not resolved yet.

veronikaslc replied:
Right, addressed here 9ed9480

marta- committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Some refactoring and code style improvements

marta- committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed buggy selection from vocabulary browser, where vocabulary branches displayed together appear to replace each other's selection rather than appending

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Addressed codereview questions

marta- commented:
> LFS mode, new Patient Information form, search for something in the Race question, more info, browse. Go up to a term that has many children, then select several of these children. Close the vocabulary browser and term info dialogs. All the selected terms appear selected. Save, close, open again -> only one of the terms was actually saved.

This bug is still present. Commit 96823e0 only fixed the UI bug in the VocabularyBrowser.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBrowser.jsx:
```diff
@@ -215,10 +214,10 @@ function VocabularyBrowser(props) {
           allowTermSelection={allowTermSelection}
           initialSelection={initialSelection}
-          questionText={questionText}
+          questionText={questionDefinition?.text}
```
Same for VocabularyTree, pass questionDefinition as a whole.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Addressed codereview questions - same optimisation for VocabularyTree component

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -43,19 +43,19 @@ import { REST_URL, MakeRequest } from "./util.jsx";
   const [ roots, setRoots ] = useState(vocabulary.roots);
+  const maxAnswers = questionDefinition?.maxAnswers;
   const addCheckbox = allowTermSelection && Number.isInteger(maxAnswers) && maxAnswers != 1;
   const addRadio = allowTermSelection && maxAnswers == 1;
```
Compute the SelectorComponent here and pass that to VocabularyBranch instead of these two props.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Code optimisation for VocabularyTree and VocabularyBranch components

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -57,7 +57,7 @@ function VocabularyTree(props) {
   const [ roots, setRoots ] = useState(vocabulary.roots);
   const maxAnswers = questionDefinition?.maxAnswers;
   const addCheckbox = allowTermSelection && Number.isInteger(maxAnswers) && maxAnswers != 1;
```
Leftover, should be removed

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -108,7 +108,16 @@ const thesaurusStyle = theme => ({
+    margin: theme.spacing(0.5, 0.5),
+  },
+  termSelector: {
+    margin: theme.spacing(-.25, 0, 0, 1.5),
```
Typo in the specs:
```suggestion
    margin: theme.spacing(-.25, 0, 0, -1.5),
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -42,14 +43,24 @@ import { REST_URL, MakeRequest } from "./util.jsx";
   const [ parentNode, setParentNode ] = useState();
   const [ currentNode, setCurrentNode ] = useState();
   const [ roots, setRoots ] = useState(vocabulary.roots);
+  const maxAnswers = questionDefinition?.maxAnswers;
+  const addCheckbox = allowTermSelection && Number.isInteger(maxAnswers) && maxAnswers != 1;
+  const selectorComponent = addCheckbox ? Checkbox : Radio;
```
Now all instances of the VocabularyBrowser have radio buttons by default (including filters and the admin page browser).

How about this:
```suggestion
  const selectorComponent = allowTermSelection ? (questionDefinition?.maxAnswers == 1 ? Radio : Checkbox) : undefined;
```
and in `VocabularyBranch` check if `selectorComponent` is defined before rendering.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -75,6 +78,12 @@ function VocabularyQuery(props) {
+  // Update a list of currently selected terms upon any interaction with the multiple choice list
+  useEffect(() => {
+    // Clear input field if maxAnswers=1
+    questionDefinition?.maxAnswers === 1 && inputValue && initialSelection && inputValue != initialSelection[0][1] && setInputValue("");
```
This doesn't look right, shouldn't it be `initialSelection[0][0]` (check the name, not the path)?

```suggestion
    questionDefinition?.maxAnswers === 1 && inputValue && initialSelection && inputValue != initialSelection[0][LABEL_POS] && setInputValue("");
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -277,6 +286,12 @@ function VocabularyQuery(props) {
+  let onCloseBrowser = (selectedTerms, removedTerms) => {
+    selectedTerms.map(item => onClick(item[VALUE_POS], item[LABEL_POS]));
+    removedTerms.map(item => removeOption(item[VALUE_POS], item[LABEL_POS]));
+    questionDefinition?.maxAnswers === 1 && setInputValue(name);
```
I'm not sure about this, what if it's one of the predefined options?

veronikaslc replied:
One way: we can check if it's predefined using `questionDefinition` and add it as a condition. Wdyt?

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Typo fix

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Better selectorComponent definition

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Fixed clear input if maxAmswers = 1

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc reviewed (commented):
/ignore, happened by accident/

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed saving selected terms bug

sashaandjic commented:
Suggestion: Since we now close the selection dialog (invoked after clicking on `i` button + `LEARN MORE`) only when user clicks on `X`, I would suggest also closing the info dialog and cleaning what users originally typed. 
The current implementation made sense when we were closing the selections if user clicked outside of the list.

![image](https://user-images.githubusercontent.com/23243244/122110135-329c1d80-cdec-11eb-8d51-c5830c176e7b.png)
![image](https://user-images.githubusercontent.com/23243244/122110080-22843e00-cdec-11eb-8680-2b880678ac05.png)

veronikaslc commented:
@sashaandjic I'm not sure about closing everything. The specs were clear that all dialogs have to remain after closing vocabulary tree. User may want still to interact with them and not loose the input and the suggestion list.

sashaandjic commented:
I can add free text by manipulating selections.
Step to reproduce:
- type "ab" in the `Race/Ethnicity reported` field
- click on any `i` button
- click on `LEARN MORE`
- make some selections
- remove all selections by clicking on the `x` button shown in the "bubble" right below the dialog title
- click on `X` to close the selection dialog
- click on the search bar next to "ab"
- ab gets selected as Race/Ethnicity reported

https://user-images.githubusercontent.com/23243244/122111302-9410bc00-cded-11eb-85b4-3c0f46789993.mp4

sashaandjic commented:
Another way to add free test is by doing this:
- type "br"
- click on `i` next to `British` option
- click on `LEARN MORE'
- select British and European
- close selection dialog
- click on the search line
- observe 3 selections: `Caucasian `(synonym for `Europian`), `British `and `"br"`

sashaandjic commented:
Synonyms are not cleared in the vocabulary tree. In the sample below, removing `Caucasian" should result in removing the `Europian` as these two are synonyms.

https://user-images.githubusercontent.com/23243244/122119775-94ae5000-cdf7-11eb-9ec8-4c87927dfd06.mp4

sashaandjic commented:
Recently selected options are not cleared from the vocabulary tree:

https://user-images.githubusercontent.com/23243244/122122055-51091580-cdfa-11eb-96b8-b0d0f59c7752.mp4

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -157,35 +157,35 @@ function MultipleChoice(props) {
-        && (!noneOfTheAboveOption || option[VALUE_POS] != noneOfTheAboveOption)
-    });
+    setSelection( old => {
+      let newSelection = old.slice().filter((option) => {
```
`filter` already creates a copy, so `slice` is probably not necessary.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +225,16 @@ function VocabularyBranch(props) {
+                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}
+                  component="div">
+        {/* Browser term select tools */}
+	    <SelectorComponent
```
Don't forget to check if `selectorComponent` is defined. Right now attempting to browse a vocabulary in the admin page or to open the vocabulary browser from the filters results in a white screen.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -145,14 +184,20 @@ function VocabularyTree(props) {
+      { allowTermSelection &&
+        <div className={classes.selectionContainer}>
+          <Typography variant="body2" component="span">{questionDefinition?.text}:</Typography>
+          { selectedTerms?.filter(i => i[LABEL_POS]).map(s => <Chip key={s[VALUE_POS]} variant="outlined" size="small" label={s[LABEL_POS]} onDelete={() => onRemoveOption(...s)} color="primary" className={classes.selectionChips}/>) }
```
Let's make this easier to read:
```suggestion
          { selectedTerms?.filter(i => i[LABEL_POS]).map(s =>
             <Chip
               key={s[VALUE_POS]}
               variant="outlined"
               size="small"
               color="primary"
               label={s[LABEL_POS]}
               onDelete={() => onRemoveOption(...s)}
               className={classes.selectionChips}
             />
           )}
```

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Synchronized terms selections and chips remove action; Code optimisation anch component existence check

veronikaslc commented:
Last two reported bugs are the same, addressed in c727f3a.

marta- commented:
Bug: The selection from the browser doesn't show up in the vocabulary question after closing the browser, but it is correctly saved. See screencast below.

![CPT2106151829-957x886](https://user-images.githubusercontent.com/651980/122131263-ead6bf80-ce06-11eb-822d-919281eb7518.gif)

veronikaslc commented:
@marta- Can not reproduce last bug

marta- commented:
Buggy behavior when switching Race/Ethnicity to maxAnswers=1:
* input value is shown as a chip when opening the browser
* any selected radio button generates a new chip intsead of replacing the existing one
* more than one radio button can appear selected at once
* clicking on a radio button focuses that term
Also, similarly to `maxAnswers != 1`, the selection from the browser is not visible in the form after closing the browser, but the correct value is saved.
 
![CPT2106151839-950x818](https://user-images.githubusercontent.com/651980/122132225-6f760d80-ce08-11eb-9dae-d8ecf4ef6dfc.gif)

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Checking if the selected term is predefined, not to set it as input value in case of single answer question

marta- commented:
> > Suggestion: Since we now close the selection dialog (invoked after clicking on `i` button + `LEARN MORE`) only when user clicks on `X`, I would suggest also closing the info dialog and cleaning what users originally typed.
> > The current implementation made sense when we were closing the selections if user clicked outside of the list.

> @sashaandjic I'm not sure about closing everything. The specs were clear that all dialogs have to remain after closing vocabulary tree. User may want still to interact with them and not loose the input and the suggestion list.

When the selection was changed from the browser, we should close everything. When nothing changed and the user just looked around (like it was happening before we introduced checkboxes and radios), we should leave everything as it was.

When `allowTermSelection` is `true`, let's remove the (x) close button and add action buttons at the bottom of the vocabulary browser:
* **Done** (primary color) - updates the selection and closes VocabularyBrowser, info box if any, suggestions dropdown, clear input if applicable (maxAnswers != 1);
* **Cancel** (default color) discards any selection done in the browser; closes VocabularyBrowser, but leaves suggestion list and input value as it was.

sashaandjic commented:
Problem with vocabulary browser when maxAnswers=1.
In the attached video I follow these steps:
- open form with `Eskimo `selection in the `Race/Ethnicity reported` field
- go to edit mode
- type br in the  `Race/Ethnicity reported` field
- click on `i` next to `British`
- click on `LEARN MORE`
- **problem #1** - notice "br" selection at the top
- **problem #2** - I can select multiple choices 
- **problem #3** - when I close the vocabulary browser I see only free text "br" as selection
- when I close and view the form I see `Dutch` which is the latest selection that I made and the correct one

https://user-images.githubusercontent.com/23243244/122140642-c506e600-ce19-11eb-9270-73f1ee8ad55e.mp4

veronikaslc commented:
@marta- @sashaandjic I think we are duplicating reviews.

marta- commented:
> @marta- @sashaandjic I think we are duplicating reviews. All bugs from the last comment are already reported.

I asked for confirmation on the two buggy behaviors I reported. The first one doesn't seem to happen anymore, the 2nd one has simply been confirmed in greater detail.

veronikaslc commented:
I see. May be it would make sense to aggregate all confirmed to one. Each review takes more than a whole screen, its taking more time to follow and concentrate if we do multiple reports with videos on same bugs, easpecially if someone else will review.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed issues with vocabulary browser when maxAnswers=1;
Added AnswerInstructions to indicate over limit selections;
Added actions for vocabulary browser for allowTermSelection is true case instead of close dialog action;
Added onClick action for selected terms chips to ficus browser of that term;

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed free term added tot he selected terms

sashaandjic commented:
DONE and CANCEL buttons look great in the vocabulary browser.
One small thing that I think should be fixed: when tabbing of the vocabulary field, clear the field

For example:
- type "p" in Race/Ethnicity field
- the list of option appears
- if you click on any other field, the selection list disappears and "p" gets removed as expected
- if you tab of the field, nothing happens (ut should behave the same way as when user clicks on any other field)

sashaandjic reviewed (commented):
Problem found:
- set maxAnswers = 1
- type `ita`
- click on `i` next to `Italian`
- select one option under `Italian`
- click on `DONE`
- expected outcome: you should see the option that you selected
- actual outcome: no option is selected
- refresh
- selected option appears


https://user-images.githubusercontent.com/23243244/122171891-39f41300-ce4e-11eb-9964-9d634baa358d.mp4


Another problem is that it is difficult to remove this option. The only way is to select one of the predefine options, save it, then open the form in edit mode and select the blank option.

sashaandjic commented:
Problem found. Steps to reproduce:
- type `"ita"` in the `Race/Ethnicity reported` field
- click on `i` next to `Italian`
- click on `LEAR MORE`
- select 5 options under Italian
- click on `DONE` --> all good
- type any letter to the `Race/Ethnicity reported` field
- click on any `i` button, then on `LEARN MORE` button
- click on the `X` buttons to remove all but one option
- click on `DONE`
- expected outcome : only one value
- actual outcome: only one option was removed
- refresh
- only one value is present as expected

https://user-images.githubusercontent.com/23243244/122169506-770ad600-ce4b-11eb-95ad-f75cd1d2bd41.mp4

marta- commented:
> Another problem is that it is difficult to remove this option. The only way is to select one of the predefine options, save it, then open the form in edit mode and select the blank option.

I confirm with simpler steps that don't involve the vocabulary browser:
* ensure maxAnswers = 1
* type `it` in the vocabulary question's input
* select Italian
* save and view
* edit
* clear the input
* save and view

Expected result: no value set. Actual result: "Italian" is still selected.

It doesn't happen in dev, so it appears to be a regression and should be fixed in this branch.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed regression of not showing answer in the input for maxAnswers=1 question case

veronikaslc commented:
Fixed regression, working on the tab issue.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -86,7 +96,7 @@ function VocabularyQuery(props) {
         delayLookup(event.target.value);
         setInputValue(event.target.value)
-        onChange && onChange(event);
+        maxAnswers === 0 && onChange && onChange(event);
```
Shouldn't this be maxAnswers != 1 ?

veronikaslc replied:
Not exactly, this causes free text to be selected for cases maxAnswers > 0. I'm adding a clause for maxAnswers == 1 and empty input in next commit.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed issue with difficulty to clear the input for single answer questions

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed options not being visually removed using chips

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Clean input anch close aoutocomplete suggestions list on tab from input

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -85,8 +95,8 @@ function VocabularyQuery(props) {
-        setInputValue(event.target.value)
-        onChange && onChange(event);
+        setInputValue(event.target.value);
+        (maxAnswers === 0 || maxAnswers === 1 && !event.target.value) && onChange && onChange(event);
```
I don't get why `onChange` needs to be called for maxAnswers = 0 or 1, but not for other values, can you explain?

veronikaslc replied:
Changing it to just `maxAnswers != 1` as you proposed earlier.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -163,6 +251,23 @@ function VocabularyTree(props) {
+          <Button color="primary"
+                  onClick={onDone}
+                  variant="contained"
+                  disabled={maxAnswers > 0 && selectedTerms.length > maxAnswers || selectedTerms.length == 0 && removedTerms.length == 0}
```
When there are values selected already from outside of the browser, when the browser is open "Done" is enabled. It should be disabled to indicate that nothing changed.

`selectedTerms` should be compared to `initialSelection` to decide if there are any changes.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -200,8 +266,17 @@ function VocabularyBranch(props) {
       {/* Term name */}
-      <Typography onClick={() => loadTerm(id, path)}
-                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}>
+      <Typography onClick={(evt) => {(evt.target.type != "checkbox") && loadTerm(id, path)}}
+                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}
+                  component="div">
+        {/* Browser term select tools */}
+	    { SelectorComponent && <SelectorComponent
+	      checked={selectedPaths.includes(path)}
+	      color="secondary"
+	      onChange={onSelectionChanged}
+	      onClick={event => event.stopPropagation()}
+	      className={classes.termSelector}
+	    /> }
```
Why not move the `SelectorComponent` outside of the `Typography` component, so that we don't need to check `evt.target.type` in the `onClick` event?
```suggestion
      {/* Browser term select tools */}
      { SelectorComponent && <SelectorComponent
          checked={selectedPaths.includes(path)}
          color="secondary"
          onChange={onSelectionChanged}
          onClick={event => event.stopPropagation()}
          className={classes.termSelector}
        /> }
      <Typography onClick={() => {loadTerm(id, path)}}
                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}
                  component="div">
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -135,24 +193,54 @@ function VocabularyTree(props) {
         onError={onError}
         knownHasChildren={!!hasChildren}
+        selectorComponent={selectorComponent}
+        addOption={onAddOption}
+        removeOption={onRemoveOption}
```
These are named a bit backwards, the handler prop should be `on<EventName>`, while the name function passed to the prop should reflect what the function does. I think the code would be easier to understand like this
```suggestion
        onTermSelected={addOption}
        onTermUnselected={removeOption}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -336,6 +338,17 @@ function MultipleChoice(props) {
               onUpdate && onUpdate(value);
             }}
+            initialSelection={selection}
+            removeOption={(value, label) => {
```
`onRemoveOption`

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -192,7 +205,7 @@ function VocabularyQuery(props) {
                   if (e.target.localName === "li") {
                     onClick(element["@path"], name);
-                    setInputValue(clearOnClick ? "" : name);
+                    setInputValue((maxAnswers !== 1 && clearOnClick) ? "" : name);
```
Why is this change needed, shouldn't the `clearOnClick` prop be sent with the proper value already?

veronikaslc replied:
It should, changing it back

marta- commented:
The following issues were found while testing in the **Multiple Choice Test** form available in the **test** runmode.

**REGRESSION**: In a vocabulary question with maxAnswers=1 and no predefined options, the input becomes disabled after setting a value and saving.
![CPT2106161656-746x840](https://user-images.githubusercontent.com/651980/122293248-d35e0c00-cec4-11eb-991d-4bb486b0f233.gif)

**BUG**: In a vocabulary question with maxAnswers=1 and no predefined options, after setting a value from the suggestions dropdown and then opening the browser, two chips with that value appear in the browser:
![CPT2106161658-733x837](https://user-images.githubusercontent.com/651980/122293311-e375eb80-cec4-11eb-94d2-1f704bb2e37d.gif)

**RE_SURFACED BUG**: In a vocabulary question with maxAnswers=1 and with predefined options, the input value is empty after making a selection from the browser, but it is saved correctly.
![CPT2106161713-711x830](https://user-images.githubusercontent.com/651980/122293938-8af31e00-cec5-11eb-95ef-35362e121046.gif)

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Small code refactoring ans better Done button disabling check

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed resurfaced bug when selected term does not appear in the input for maxAnswers=1 case

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed single answer question without default answers

veronikaslc commented:
The regression bug and the one after it with double chips have the same origin. MultipleChoice was adding selected term twice which locked the input because of selected terms number bigger than 1.

sashaandjic commented:
DONE button should always be enabled in the Vocabulary browser. Currently, DONE button gets enabled only if user selects a new option. Sometimes user would just want to delete one or more and save it, which is impossible right now.

https://user-images.githubusercontent.com/23243244/122349019-cddee100-cf19-11eb-8cb7-2c1de52444e9.mp4

sashaandjic commented:
While testing tabbing, I noticed that I can no longer land on the Family field either by tabbing from the field above nor by clicking on it.

sashaandjic commented:
- set maxAnswers=1
- select one of the predefine options in Race/Ethnicity
- save and close
- open in edit mode
- the text of the predefined option is copied at the bottom

https://user-images.githubusercontent.com/23243244/122352414-1c41af00-cf1d-11eb-8122-bc77ef4120bc.mp4

veronikaslc commented:
All above is by design.
Done button disabled by request:
>When there are values selected already from outside of the browser, when the browser is open "Done" is enabled. It should be disabled to indicate that nothing changed

Single answer input is filled with selected option to indicate that it's selected, if input is empty nothing is selected as the answer. 

Is Family a vocabulary question in some of other forms? If not - its out of the scope of this PR.

sashaandjic commented:
I would  enable `DONE` button in the browser if user removes the selection by pressing on `x`, as shown in the video.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Enable Done button if terms are removed from the vocabulary tree

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -163,6 +251,25 @@ function VocabularyTree(props) {
+                  onClick={onDone}
+                  variant="contained"
+                  disabled={maxAnswers > 0 && selectedTerms.length > maxAnswers
+                         || removedTerms.length == 0
+                         || selectedTerms.filter(([a1,a2]) => !initialSelection.some(([s1,s2]) => a2 === s2)).length == 0}
```
Now the button is always disabled when maxAnswers=1.

```suggestion
                         || removedTerms.length == 0
                            && selectedTerms.every(([a1,a2]) => initialSelection.find(([s1,s2]) => a2 === s2))}
```

marta- commented:
> All above is by design. Shell we change? @marta-
> Done button disabled by request:
> 
> > When there are values selected already from outside of the browser, when the browser is open "Done" is enabled. It should be disabled to indicate that nothing changed

I guess the request should have been more explicit and continued with: "as soon as there is any change, either new term selected or a selection is removed, the Done button should be enabled to allow for saving the change."

> 
> Single answer input is filled with selected option to indicate that it's selected, if input is empty nothing is selected as the answer.

I disagree. If there are predefined options and one of them is selected, there's no need to fill the input with the same value, it looks like a bug.

> 
> Is Family a vocabulary question in some of other forms? If not - its out of the scope of this PR.

It might not be completely out of scope. The "Family" question is right after a vocabulary question (Race/Ethnicity in the Patient information form). As soon as a value is added for Race/Ethnicity, we can no longer tab into Family. This doesn't happen in the `dev` branch. It should at least be investigated further.

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Enable Done button for cases when maxAnswers=1.

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

marta- commented:
**BUG**: in vocabulary questions **without** predefined options, selecting a term from the browser does not populate the input with the term's label, [see screencast](https://user-images.githubusercontent.com/651980/122478863-28b71d80-cf98-11eb-9dd9-2a86005d4bd6.gif). This does not happen when there are predefined options.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Show single answer in the input only if it is not one of defaults

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed bug: in vocabulary questions without predefined options, selecting a term from the browser does not populate the input with the term's label

veronikaslc commented:
>It might not be completely out of scope. The "Family" question is right after a vocabulary question (Race/Ethnicity in the Patient information form). As soon as a value is added for Race/Ethnicity, we can no longer tab into Family. This doesn't happen in the dev branch. It should at least be investigated further.

Can not reproduce it.

sashaandjic commented:
When maxAns=1, I should be able to clear option by using vocab browser (like I do when maxAns!=1)

[Unable to clear](https://user-images.githubusercontent.com/23243244/122486241-a84be900-cfa6-11eb-9493-5e1ff47431a0.mp4
)

sashaandjic commented:
Are we going to fix this behaviour when maxAns=1?

https://user-images.githubusercontent.com/23243244/122486360-ecd78480-cfa6-11eb-9528-a1e97779a5dc.mp4

sashaandjic commented:
Weird behaviour when maxAns=2:
- set maxAnswers=2
- type something in ethnicity field and open the browser ( `i` + `LEARN MORE`)
- select one option --> close browser --> all good
- remove selected option in the form
- repeat ( type something, open browser, select one option, close browser)
- free text appears as selection

https://user-images.githubusercontent.com/23243244/122486653-830baa80-cfa7-11eb-9653-51db70319c58.mp4

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -98,7 +98,7 @@ function MultipleChoice(props) {
   const [ghostValue, setGhostValue] = useState(inputPrefill?.value || GHOST_SENTINEL);
   const ghostSelected = selection.some(element => {return element.includes(ghostValue);});
-  const disabled = maxAnswers > 0 && selection.length >= maxAnswers && !isRadio && !ghostSelected;
+  const disabled = maxAnswers > 0 && selection.length >= maxAnswers && maxAnswers !== 1 && !ghostSelected;
```
I think this can be simplified as:
```suggestion
  const disabled = maxAnswers > 1 && selection.length >= maxAnswers;
```

sashaandjic commented:
@veronike here are the steps to see the problem with tabbing:
- select some option in the `Race/Ethnicity reported` field
- start tabbing
- once you tab off the `Family `field, the field becomes disabled and you can no longer land on it

https://user-images.githubusercontent.com/23243244/122487409-31fcb600-cfa9-11eb-9785-c3c7766edbf4.mp4

veronikaslc commented:
> Are we going to fix this behaviour when maxAns=1?
Already fixed

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
clear option by using vocab browser

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Do not trigger option selection on empty input on input focus

sashaandjic commented:
Strange ui behaviour where predefined options appear to be removed.

https://user-images.githubusercontent.com/23243244/122504768-f07c0300-cfc8-11eb-8769-a4620f33ac96.mp4

sashaandjic commented:
All previously reported problems are fixed.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -48,20 +49,30 @@ const MAX_RESULTS = 10;
+  let isDefaultOption = (path) => {
+    return Object.values(props.questionDefinition)
+          .filter(value => value['jcr:primaryType'] == 'lfs:AnswerOption' && value.value === path)
+          .length > 0;
```
I find this easier to read:
```suggestion
          .find(value => value['jcr:primaryType'] == 'lfs:AnswerOption' && value.value === path)
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -277,6 +298,29 @@ function VocabularyQuery(props) {
+      } else {
+        if (initialSelection?.length > 0) {
+          !isDefaultOption(initialSelection[0][VALUE_POS]) && setInputValue(initialSelection[0][LABEL_POS]);
+        }
```
This `else` branch should be removed. When clicking cancel in the browser, everything should be left as it was before opening the browser. Instead, this resets to the initial value.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -163,6 +251,25 @@ function VocabularyTree(props) {
+          <Button color="primary"
+                  onClick={onDone}
+                  variant="contained"
+                  disabled={maxAnswers > 0 && selectedTerms.length > maxAnswers
+                         || removedTerms.length == 0
+                            && selectedTerms.every(([a1,a2]) => initialSelection.find(([s1,s2]) => a2 === s2))}
```
To avoid the user's confusion when they click around the browser changing the selection only to end up with a disabled "Done" button because their current selection happens to be identical to the initial selection, let's do the folowing:
* add another state variable `selectionChanged`, initially `false`
* add an `effect` to set `selectionChanged` to `true`  when `selectedTerms` or `removedTerms` change
* make `disabled={!selectionChanged}` here

veronikaslc replied:
What about `maxAnswers > 0 && selectedTerms.length > maxAnswers` condition of diabling the button?

marta- replied:
Right, sorry, I overlooked that part. Should be `disabled={maxAnswers > 0 && selectedTerms.length > maxAnswers || !selectionChanged}`.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -108,7 +108,20 @@ const thesaurusStyle = theme => ({
     zIndex: "1350 !important",
-  }
+  },
+  selectionContainer: {
+    padding: theme.spacing(0, 3, 2),
+  },
+  browserAnswerInstrustions: {
+    padding: theme.spacing(1, 3, 0),
+    marginBottom: theme.spacing(-2),
+  },
+  selectionChips: {
+    margin: theme.spacing(0.5, 0.5),
+  },
+  termSelector: {
+    margin: theme.spacing(-.25, 0, 0, -1.5),
+  },
```
Aren't all these related to the VocabularyBrowser? Shouldn't they be in `browseStyle` instead?

veronikaslc replied:
They should, moving them there.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -332,10 +334,13 @@ function MultipleChoice(props) {
               onUpdate && onUpdate(value);
             }}
+            initialSelection={selection.filter(option => option[VALUE_POS])}
+            onRemoveOption={(value, label) => {
+              removeOption(value, label);
+            }}
```
This may be the cause of the bug with disappearing default options encountered by @sashaandjic . `removeOption` actually removes the entry from the option list (where option list = list of default options + any new options added by the user via the input). See fix suggestions in the comments on `removeOption`.

Also, this can now be simplified:
```suggestion
            onRemoveOption={removeOption}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -209,19 +209,21 @@ function MultipleChoice(props) {
     onChange && onChange(id); // will trigger callback in Form.jsx
-    setOptions(options.filter(
-      (option) => {
-        return !(option[VALUE_POS] === id && option[LABEL_POS] === name)
-      }
-    ));
+    setOptions( (old) => {
+      let newOptions = old.filter(
+        (option) => {
+          return !(option[VALUE_POS] === id && option[LABEL_POS] === name)
+        });
+      return newOptions;
+    });
     unselect(id, name);
     return;
```
This function is for removing non-default options. If called for a default option, it actually removes the default entry from the option list, but this should never happen in practice.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -209,19 +209,21 @@ function MultipleChoice(props) {
+    setOptions( (old) => {
+      let newOptions = old.filter(
+        (option) => {
+          return !(option[VALUE_POS] === id && option[LABEL_POS] === name)
```
This will probably prevent the removal of default options (not tested):
```suggestion
          return !(option[VALUE_POS] === id && option[LABEL_POS] === name) || option[IS_DEFAULT_POS]
```

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Simplified disabled propery calculation in MultipleChoice.jsx

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

More compact code

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser

Code optimisation

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Addressed codereview comments

sashaandjic commented:
Problem when maxAnswers=1 and user pick an otion that in not predefined:
- set maxAnswers=1
- select `Peruvian `in the `Race/Ethnicity reported` field
- save and close
- open in edit mode, option `Peruvian` appears at the bottom but it is not selected

https://user-images.githubusercontent.com/23243244/122825687-434b0880-d2b0-11eb-819f-8fa7848d49fb.mp4

I would like to see the same behaviour as when selecting predefined option, as shown in the video below:

https://user-images.githubusercontent.com/23243244/122825765-5bbb2300-d2b0-11eb-8fe6-a15d7498d90d.mp4

veronikaslc commented:
Can not reproduce the above problem. @marta- do you observe the same?

sashaandjic commented:
The strange thing is that I can no longer reproduce it in the same instance with new subjects. However, the first patient/form still has this problem.

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed Problem when maxAnswers=1 and user pick an option that is not predefined

veronikaslc commented:
@sashaandjic should be fixed now

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -60,6 +67,33 @@ function VocabularyBranch(props) {
+  let removePath = (evt) => {
+    let path = evt.detail[VALUE_POS];
+    setSelectedPaths(old => {
+        let newPaths = old.filter(item => item!= path);
```
```suggestion
        let newPaths = old.filter(item => item != path);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -169,6 +208,33 @@ function VocabularyBranch(props) {
+        onTermSelected(name, path);
+      }
+    } else {
+      let newPaths = selectedPaths.filter(item => item!= path);
```
```suggestion
      let newPaths = selectedPaths.filter(item => item != path);
```

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed info buttons refs array to support same terms ref from multiple parents

Commit:
[misc] Better code style; Removed unused import

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed duplicate selection of terms with multiple parents

veronikaslc commented:
Rebased on the latest dev to include all other fixes.

sashaandjic commented:
@sdumitriu reported the following 3 issues:
-  install `HP`, tumors form, search for `obesity`, info on `Truncal obesity`, Learn more, select `Childhood-onset truncal obesity`, navigate up to `Obesity`, select `Obesity`, expand `Truncal obesity` -> both `obesity `and `childhood are selected`
- search for and select `Obesity `in Diagnosis code, save the form (don't close), delete the text in Diagnosis code, focus out -> field seems empty. Save -> `Obesity` pops back in
-  search `Obesity`, info on obesity, Learn more, info on `Truncal Obesity` in the vocab browser, press `Esc` key -> Truncal obesity info dialog moves in the corner; moreover, pressing the info button on `Obesity `in the search results no longer works (also instead of clicking on `Esc`, click on `CANCEL `--> info button on `Obesity` no longer works)

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Added the escape button action to close popups

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Save empty input for maxAnswers = 1 case

Commit:
LFS-328: As a user answering a vocabulary question, I can select an option from the Vocabulary browser
Fixed multiple term selection in vocabulary tree for maxAnswers = 1 case

sashaandjic reviewed (approved):
tested

sashaandjic commented:
why am I not seeing any of these in the Tumor form?
![image](https://user-images.githubusercontent.com/23243244/123867118-f09acd80-d8fb-11eb-9567-51f21cc8c884.png)

sashaandjic commented:
Ok. I see, on the form we filter ORDO to show only `Orphanet_C001` term. Is this as per requirement? Last time this filter was not there.

marta- commented:
> Ok. I see, on the form we filter ORDO to show only `Orphanet_C001` term. Is this as per requirement? Last time this filter was not there.

Nobody required it specifically, but metadata such as annual incidence and prevalence don't make sense as values for "diagnosis".

sashaandjic commented:
Sorry my bad. I see that @sergiu removed the filter prior to testing and finding things reported on Slack.
